### PR TITLE
NF: Adding typing to tasks and moving code where it logically fits

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
As you know, I set myself the goal of improving Background Tasks. I sent a first PR last week. The more I thought about it, the less I liked it. It was a big PR with a single commit, it made everything hard to do: proof reading, rebasing in case of conflict... With 92 commits, here it'll paradoxally get simpler.

As I did with Undoable in another PR, I put the code near the place where it is actually used. I kept methods in CollectionTask only when they are used in multiple classes. This ensure that when you want to understand some action, you have all relevant code together.

I did split the methods form Dismiss and DismissMulti, because they were too big and hard to read. Worse, they used different type of argument and returned different type of values, which made typing complex.

And this is the second part I love. Now that there is a lot of distinct methods, everything can get typed properly ! Both input, output and updates ! I was able to remove TaskData and I really believe it'll make code easier to read and write, since the function arguments and return value will be known thanks to the type ! Typing is not entirely perfect yet; and I'm sure that some improvement will be possible later. 

The only problem I see is that every single PR dealing with background task will need to be rewritten. And given my PR queue, I'll have quite some work.

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java [hunk 1]
```diff
@@ -206,8 +229,16 @@ public boolean commit() {
                                 int oldValue = mOptions.getJSONObject("new").getInt("order");
                                 if (oldValue != newValue) {
                                     mOptions.getJSONObject("new").put("order", newValue);
-                                    CollectionTask.launchCollectionTask(REORDER, mConfChangeHandler,
-                                            new TaskData(new Object[] {mOptions}));
+                                    DeckConfig conf = mOptions;
+                                    CollectionTask.TaskCode<Void, Boolean> code = new CollectionTask.TaskCode<Void, Boolean>() {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:233]: Is there a warning around here? This appears to leak the `Activity` object

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:233]: I am failed to see how it capture the whole activity. I thought it was capturing only the elements it uses, and it does not seems to uses `this`.

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:233]: Right. I didn't know it was capturing even without accessing it. I'll need to create all objects using static values

**Comment by Arthur-Milchior** [comment]
Rewritten to use an interface instead of an abstract class, using lambda expression and moving classes to other file so that lines (types) get shorter and avoid a lot of "CollectionTask." repetition

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java]: On fourth thought, it was a problem with listener, and doing the same error with tasks was not a good idea. So instead I corrected listeners https://github.com/ankidroid/Anki-Android/pull/6775

**Comment by Arthur-Milchior** [comment]
All classes are static now. No more leak

**Comment by Arthur-Milchior** [comment]
Fun fact: while I was resolving conflict with master, more conficts appeared

**Comment by Arthur-Milchior** [comment]
A lot of conflict removed

**Comment by mikehardy** [comment]
This one was blocked by the listener-no-leak refactor which I just merged, and in turn seems to block #6741 so it is a logical next step to de-conflict and work through

**Comment by Arthur-Milchior** [comment]
#6741 will disappear if this PR is merged. #6741 goal is too make this one slightly simpler

**Comment by Arthur-Milchior** [comment]
Conflict removed again. 

I believe that this is the best PR I can make out of this idea. Splitting the huge function into something that is nicely typed and where the code is moved where we actually want it instead of being in a single class only because it's background code.
However, I could remove the typing part; it would made the PR shorter if you prefer and still interesting. Please let me know your preference.

Normally, each commit is simple to understand. It's going to take time to go through it, but if you want to ensure it's NF I believe that reading each commit will be quite simpler than reading the whole diff

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
My main concern is that we've moved a lot of business logic into the view layer. This will encourage future developers to use bad abstractions.

As a second consideration, we should use a `Result<T>` structure, or similar. 

`Pair` and `Triple` is being used here because Java's very verbose to create a record class, but it's not an ideal design. Maybe this is where we draw the line and say we want a dependency, or change in code style be it Lombok or Kotlin so we can improve this aspect of the code quality.

After the above are considered, we'll need to work out where to cut this up to make it reviewable. It's not viable to review such a big change which touches so many components right now.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java [hunk 2]
```diff
@@ -3721,13 +3941,13 @@ public int ankiGetETA() {
         }
 
         @JavascriptInterface
-        public boolean ankiGetCardMark() {
+        public boolean ankiCardMark() {
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java]: Corrected. I had to add a GetCard class which I was able to delete later. I didn't realize that there was "GetCard" outside of the class name

**Comment by Arthur-Milchior** [comment]
> My main concern is that we've moved a lot of business logic into the view layer. This will encourage future developers to use bad abstractions.

If I trust my experience, the current design does not really help understand either how to make a background tasks. I had plenty of wrong start before understanding what I should do. My intuition would tell it would give code that one can copy paste to create a task in more part of the codebase and make it easier to find. I would suspect it increases the chance future dev's do things correctly.

I think I can understand why, for example, suspending a bunch of card should not be in the view layer but in the business layer. I would be happy to have this kind of class in libanki, it may makes far more sens. Except that we want to follow upstream.


> As a second consideration, we should use a `Result<T>` structure, or similar.

Where would you want us to use it ? It does not seems straightforward to me.

> `Pair` and `Triple` is being used here because Java's very verbose to create a record class, but it's not an ideal design. Maybe this is where we draw the line and say we want a dependency (Lombok or Kotlin), so we can improve this aspect of the code quality. @mikehardy ?

I do agree that it's really not idea and a simple struct would be nice. I sincerely believes it's still far better than task data or Object[]. Especially in the card browser where it was really hard to follow the logic of data passed between task and listener. 


> 
> After the above are considered, we'll need to work out where to cut this up to make it reviewable. It's not viable to review such a big change which touches so many components right now.

There is an obvious place to cut, before typing. The place where the code is moved where it logically belongs is sufficient in itself.

Ideally I would love to cut more, but that would not be practical, because I'm changing the whole logic by introducing the class `Task`, and it would not make sens to have some tasks using `Task` and some using the huge collection task switch (and that would be a nightmare to merge)

**Comment by Arthur-Milchior** [comment]
I'd like to argue that this PR is far more important than what I expected. And I may be able to convince you to consider it seriously.

One of the current problem I've had with doing realistic test is that they all depends on background tasks. It's extremely hard not to depend on background tasks because the same class `CollectionTask` deals with processing the task queue and doing the actual background job.

If at some point you accept this PR, all methods will be in separate `Task` classes. It will then be really really easy to call a task, which normally should be called via `launchCollection`, in the main thread. This means that tests does not have to deal with executing background tasks and checking whether a tasks did indeed execute.

For example, in the card browser, it would allow me to test deletion easily, since I would just need to call the delete function and would not need to check whether task ended or not to check the expected result.

Of course, I admit that this PR is not sufficient by itself. The missing step is to create a mock of CollectionTask. This mock would simply transmit calls directly between caller, listener and tasks without using any creation of actual task. That would become easy to do. However, I'd prefer for us to wait to have `Task` in the codebase before I start to add more commits to this PR

**Comment by Arthur-Milchior** [comment]
I still believe that CollectionTask is uselessly complex and should be changed. At least to use proper typing instead of TaskData. And also that https://github.com/ankidroid/Anki-Android/pull/7046 is a simpler first step, by differentiating the manager from the tasks

**Comment by david-allison-1** [comment]
Looking at this with fresh eyes, the architectural problem is that `CollectionTask` acted as a "Model" layer in the project, there was clear separation between Activity/UI logic and business logic.

With this change, we lose the distinction. It'd be better if we could keep this model layer, but I'm not quite sure about the design yet: one class per task seems unwieldy. Likely a "class per concept" would work best.

**Comment by Arthur-Milchior** [comment]
What do you mean by "concept" ?

**Comment by david-allison-1** [comment]
> What do you mean by "concept" ?

Scheduling, Browsing, Editing etc... - maybe this is a 1:1 mapping with the current classes in libAnki

**Comment by Arthur-Milchior** [comment]
This was splitted into:
* https://github.com/ankidroid/Anki-Android/pull/7046 , 
* https://github.com/ankidroid/Anki-Android/pull/7685
* https://github.com/ankidroid/Anki-Android/pull/7562

I guess I'll close it. Even if at some point we agree about moving tasks in different file based on their meaning, I believe that this PR will be unusable assuming the three other PR are accepted
